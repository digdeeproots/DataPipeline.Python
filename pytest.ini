[pytest]
asyncio_mode=auto
pythonpath = src
testpaths = tests
python_files = *.py
norecursedirs = .pytest_cache .github .idea
addopts = --showlocals --import-mode=importlib
