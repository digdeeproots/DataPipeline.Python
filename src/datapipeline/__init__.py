from __future__ import annotations

from abc import ABCMeta, abstractmethod
from typing import TypeVar, Generic, runtime_checkable, Protocol

T = TypeVar('T')
U = TypeVar('T')
TIn = TypeVar('TIn')
TOut = TypeVar('TOut')


@runtime_checkable
class NamedStep(Protocol):
    @property
    @abstractmethod
    def name(self) -> str:
        pass


@runtime_checkable
class ProcessingStep(NamedStep, Protocol[TIn]):
    @abstractmethod
    def transform(self, data: TIn) -> None:
        pass


@runtime_checkable
class RestructuringStep(NamedStep, Protocol[TIn, TOut]):
    @abstractmethod
    def restructure(self, data: TIn) -> TOut:
        pass


class _PipeSegment(Generic[TIn, TOut], metaclass=ABCMeta):
    _next_segment: _PipeSegment[TOut, U]

    def __init__(self):
        self._next_segment = NullTerminator()

    @property
    @abstractmethod
    def name(self) -> str:
        pass

    def to_verification_string(self) -> str:
        return f'+--{self.name}'

    @abstractmethod
    def _process(self, data: TIn) -> TOut:
        pass

    def process(self, data: TIn) -> None:
        result = self._process(data)
        self._next_segment.process(result)

    def then(self, next_segment: ProcessingStep[TOut] | RestructuringStep[TOut, U]) -> _PipeSegment[TOut, U]:
        self._next_segment = DataProcessingSegment(next_segment) if isinstance(next_segment, ProcessingStep) else \
            RestructuringSegment(next_segment)
        return self._next_segment, next_segment


class DataProcessingSegment(_PipeSegment[T, T], Generic[T]):
    _impl: ProcessingStep[T]

    def __init__(self, impl: ProcessingStep[T]):
        super(DataProcessingSegment, self).__init__()
        self._impl = impl

    @property
    def name(self) -> str:
        return self._impl.name

    def _process(self, data: TIn) -> TOut:
        self._impl.transform(data)
        return data


class RestructuringSegment(_PipeSegment[TIn, TOut], Generic[TIn, TOut]):
    _impl: RestructuringStep[TIn, TOut]

    def __init__(self, impl: RestructuringStep[TIn, TOut]):
        super(RestructuringSegment, self).__init__()
        self._impl = impl

    @property
    def name(self) -> str:
        return self._impl.name

    def _process(self, data: TIn) -> TOut:
        return self._impl.restructure(data)


class NullTerminator(_PipeSegment[T, T], Generic[T]):
    def __init__(self):
        pass  # Don't call the superclass.

    def name(self) -> str:
        return "End"

    def process(self, data: TIn) -> None:
        pass

    def _process(self, data: TIn) -> TOut:
        raise NotImplementedError

    def then(self, next_segment: ProcessingStep[T] | RestructuringStep[T, U]) -> _PipeSegment[T, U]:
        raise NotImplementedError
