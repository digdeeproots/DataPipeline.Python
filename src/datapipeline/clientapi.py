from __future__ import annotations

from abc import abstractmethod
from typing import runtime_checkable, Protocol, TypeVar

TIn = TypeVar('TIn')
TOut = TypeVar('TOut')


@runtime_checkable
class NamedStep(Protocol):
    @property
    @abstractmethod
    def name(self) -> str:
        pass


@runtime_checkable
class ProcessingStep(NamedStep, Protocol[TIn]):
    @abstractmethod
    def transform(self, data: TIn) -> None:
        pass


@runtime_checkable
class RestructuringStep(NamedStep, Protocol[TIn, TOut]):
    @abstractmethod
    def restructure(self, data: TIn) -> TOut:
        pass
