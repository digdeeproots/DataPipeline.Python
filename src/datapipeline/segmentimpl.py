from __future__ import annotations

import inspect
from abc import ABCMeta, abstractmethod
from typing import Generic, TypeVar, Callable, List

from datapipeline import clientapi

U = TypeVar('U')
TIn = TypeVar('TIn')
TOut = TypeVar('TOut')
TRaw = TypeVar('TRaw')


class _PipeSegment(Generic[TIn, TOut], metaclass=ABCMeta):
    _next_segment: _PipeSegment[TOut, U]
    needs: List[str]
    gives: List[str]

    class Iterator:
        next_segment: _PipeSegment

        def __init__(self, next_segment: _PipeSegment):
            self.next_segment = next_segment

        def __iter__(self):
            return self

        def __next__(self):
            if isinstance(self.next_segment, NullTerminator):
                raise StopIteration()
            self.next_segment = self.next_segment._next_segment
            return self.next_segment

    def __init__(self, needs: List[str], gives: List[str], next_segment: _PipeSegment[TOut, U] = None):
        if next_segment is None:
            next_segment = NullTerminator()
        self._next_segment = next_segment
        self.needs = needs
        self.gives = gives

    def __iter__(self):
        return _PipeSegment.Iterator(self)

    @property
    @abstractmethod
    def descriptor(self) -> str:
        raise NotImplementedError

    def to_verification_string(self) -> str:
        return f'  |\n{self.descriptor}\n  |     info change: {self.needs} ==> {self.gives}\n{self._next_segment.to_verification_string()}'

    @abstractmethod
    def _process(self, data: TIn) -> TOut:
        pass

    def process(self, data: TIn) -> None:
        result = self._process(data)
        self._next_segment.process(result)


class PipeHeadSegment(_PipeSegment[TIn, TIn], Generic[TIn]):
    _impl: Callable[[], TIn]

    def __init__(self, impl: Callable[[], TIn], next_segment: _PipeSegment[TIn, U] = None):
        super(PipeHeadSegment, self).__init__([], [], next_segment)
        self._impl = impl

    def to_verification_string(self) -> str:
        return f' <!> Start with empty <{self._impl.__name__}>\n{self._next_segment.to_verification_string()}'

    def _process(self, data: TIn) -> TIn:
        return self._impl()

    @property
    def descriptor(self) -> str:
        raise NotImplementedError


class DataProcessingSegment(_PipeSegment[TIn, TIn], Generic[TIn]):
    _impl: clientapi.ProcessingStep[TIn]

    def __init__(self, impl: clientapi.ProcessingStep[TIn], next_segment: _PipeSegment[TIn, U] = None):
        assert isinstance(impl, clientapi.ProcessingStep)
        super(DataProcessingSegment, self).__init__([], [], next_segment)
        self._impl = impl

    @property
    def descriptor(self) -> str:
        return f'{self.symbol()} {self._impl.__name__}'

    @abstractmethod
    def symbol(self) -> str:
        raise NotImplementedError

    def _process(self, data: TIn) -> TIn:
        self._impl(data)
        return data


class SourceSegment(DataProcessingSegment[TIn], Generic[TIn, TRaw]):
    def __init__(self, load: clientapi.Loader[TIn, TRaw], parse: clientapi.ParseImpl[TIn, TRaw],
                 next_segment: _PipeSegment[TIn, U] = None):
        def impl(data: TIn) -> None:
            parse(data, load(data))

        assert isinstance(load, clientapi.Loader)
        assert isinstance(parse, clientapi.ParseImpl)
        impl.__name__ = f'load:{load.__name__}, parse: {parse.__name__}'
        super(SourceSegment, self).__init__(impl, next_segment)

    def symbol(self) -> str:
        return ">-|  "


class TransformSegment(DataProcessingSegment[TIn], Generic[TIn]):
    def symbol(self) -> str:
        return "  +  "


class SinkSegment(DataProcessingSegment[TIn], Generic[TIn, TRaw]):
    def __init__(self, extract: clientapi.Extractor[TIn, TRaw], store: clientapi.StoreImpl[TRaw],
                 next_segment: _PipeSegment[TIn, U] = None):
        def impl(data: TIn) -> None:
            store(extract(data))

        assert isinstance(extract, clientapi.Extractor)
        assert isinstance(store, clientapi.StoreImpl)
        impl.__name__ = f'extract: {extract.__name__}, store: {store.__name__}'
        super(SinkSegment, self).__init__(impl, next_segment)

    def symbol(self) -> str:
        return "  |->"


class RestructuringSegment(_PipeSegment[TIn, TOut], Generic[TIn, TOut]):
    _impl: Callable[[TIn], TOut]

    def __init__(self, impl: Callable[[TIn], TOut], next_segment: _PipeSegment[TOut, U] = None):
        assert isinstance(impl, clientapi.RestructuringStep)
        super(RestructuringSegment, self).__init__([], [], next_segment)
        self._impl = impl

    @property
    def descriptor(self) -> str:
        return f' <->  Changed to <{inspect.get_annotations(self._impl, eval_str=True)["return"].__name__}> using {self._impl.__name__}'

    def _process(self, data: TIn) -> TOut:
        return self._impl(data)


class NullTerminator(_PipeSegment[TIn, TIn], Generic[TIn]):
    def __init__(self):
        pass  # Don't call the superclass.

    def __next__(self):
        raise StopIteration()

    def descriptor(self) -> str:
        raise NotImplementedError

    def to_verification_string(self) -> str:
        return ""

    def process(self, data: TIn) -> None:
        pass

    def _process(self, data: TIn) -> TIn:
        raise NotImplementedError
