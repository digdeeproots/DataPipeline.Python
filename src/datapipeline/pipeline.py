from __future__ import annotations

import asyncio
import datetime
from typing import Callable, TypeVar, Awaitable, Generic, List, Iterable

from assertpy import soft_assertions

from datapipeline import DataProcessingSegment, RestructuringSegment, PipeHeadSegment, clientapi
from datapipeline.clientapi import AfterError
from datapipeline.segmentimpl import _PipeSegment, SourceSegment, TransformSegment, SinkSegment

T = TypeVar("T")
TRaw = TypeVar("TRaw")
TSrc = TypeVar("TSrc")
TNext = TypeVar("TNext")
TDest = TypeVar("TDest")


def needs(*info_items: str) -> Callable[[T], T]:
    def inner(f: T) -> T:
        f._p_needs_ = list(info_items)
        return f

    return inner


def gives(*info_items: str) -> Callable[[T], T]:
    def inner(f: T) -> T:
        f._p_gives_ = list(info_items)
        return f

    return inner


SegmentBuilder = Callable[[_PipeSegment[TNext, T] | None], DataProcessingSegment[TNext]]
RestructureBuilder = Callable[[_PipeSegment[TNext, T] | None], RestructuringSegment[TSrc, TNext]]
PipeHeadBuilder = Callable[[_PipeSegment[TNext, T] | None], PipeHeadSegment[TNext]]
AnyBuilder = Callable[[_PipeSegment | None], _PipeSegment]


def start_with(data_constructor: Callable[[], T]) -> IncompletePipeline[T, T]:
    def build(next_segment: _PipeSegment[T, TNext] | None) -> PipeHeadSegment[T]:
        return PipeHeadSegment(data_constructor, next_segment)

    return IncompletePipeline[T, T]([build])


SomeException = TypeVar('Exc', bound=Exception)


class ErrorHandler(Generic[T, SomeException]):
    exception: type[SomeException]
    handler: clientapi.ErrorResponse[T, SomeException]

    def __init__(self, exception: type[SomeException], handler: clientapi.ErrorResponse[T, SomeException]) -> None:
        self.exception = exception
        self.handler = handler

    def __call__(self, next_segment: _PipeSegment[T, TNext] | None) -> SourceSegment[T, TRaw]:
        return next_segment


def on_error(exception: type[SomeException],
             handler: clientapi.ErrorResponse[T, SomeException]) -> ErrorHandler[T, SomeException]:
    return ErrorHandler(exception, handler)


def source(load: Callable[[T], Awaitable[TRaw]],
           parse: Callable[[T, TRaw], None],
           *error_handlers: ErrorHandler) -> SegmentBuilder[T]:
    def build(next_segment: _PipeSegment[T, TNext] | None) -> SourceSegment[T, TRaw]:
        return SourceSegment(load, parse, next_segment)

    return build


def transform(process: Callable[[T], None], *error_handlers: ErrorHandler) -> SegmentBuilder[T]:
    def build(next_segment: _PipeSegment[T, TNext] | None) -> TransformSegment[T]:
        return TransformSegment(process, next_segment)

    return build


def sink(extract: Callable[[TSrc], TDest],
         store: Callable[[TDest], Awaitable[None]],
         *error_handlers: ErrorHandler) -> SegmentBuilder[TSrc]:
    def build(next_segment: _PipeSegment[T, TNext] | None) -> SinkSegment[T, TRaw]:
        return SinkSegment(extract, store, next_segment)

    return build


class IncompletePipeline(Generic[TSrc, T]):
    def __init__(self, prior_steps: List[AnyBuilder]):
        self._prior_steps = prior_steps

    def then(self, *steps: SegmentBuilder[T] | ErrorHandler) -> PotentiallyCompletePipeline[TSrc, T]:
        # Handle error handlers. Require them to only be at the end; they apply to the whole chain.
        return PotentiallyCompletePipeline(self._prior_steps + list(steps))


class PotentiallyCompletePipeline(Generic[TSrc, T]):
    def __init__(self, prior_steps: List[AnyBuilder]):
        self._prior_steps = prior_steps

    def restructure_to(self,
                       data_constructor: Callable[[], TDest],
                       restructure: Callable[[T], TDest]) \
            -> IncompletePipeline[TSrc, TDest]:
        def build(next_segment: _PipeSegment[T, TNext] | None) -> RestructuringSegment[T, TDest]:
            return RestructuringSegment(restructure, next_segment)

        return IncompletePipeline(self._prior_steps + [build])

    def build(self) -> Pipeline:
        next_segment = None
        for builder in reversed(self._prior_steps):
            next_segment = builder(next_segment)
        return Pipeline(next_segment)


class Pipeline:
    _first_segment: _PipeSegment

    def __init__(self, first_segment):
        self._first_segment = first_segment

    def to_verification_string(self):
        return self._first_segment.to_verification_string()

    @property
    def _segments(self) -> Iterable[_PipeSegment]:
        return iter(self._first_segment)

    def run(self):
        asyncio.run(self._first_segment.process(None))


def pipeline(builder: PotentiallyCompletePipeline[TSrc, TDest], *error_handlers: ErrorHandler) -> Pipeline:
    return builder.build()


def retry(max_retries: int, delay: datetime.timedelta) -> clientapi.ErrorResponse:
    retries_left = max_retries

    async def do_retry(
            segment: _PipeSegment[T, TNext],
            original_data: T,
            modified_data: T,
            exception: Exception) -> clientapi.AfterError:
        nonlocal retries_left
        retries_left -= 1
        if retries_left > 0:
            await asyncio.sleep(delay.seconds)
            return clientapi.AfterError.Retry

    return do_retry


def skip_this_step(
        segment: _PipeSegment[T, TNext],
        original_data: T,
        modified_data: T,
        exception: Exception) -> AfterError:
    return AfterError.Skip


def is_valid_pipeline(self):
    if not isinstance(self.val, Pipeline):
        raise TypeError('val must be a pipeline.')
    provided = set()
    with soft_assertions():
        kind = self.kind
        self.kind = 'soft'
        for segment in self.val._segments:
            needed = set(segment.needs)
            if not needed <= provided:
                desc = self.description
                self.description = f'Segment {segment.descriptor}'
                self.error(f'Had unmet needs {needed - provided}. The pipeline to that point had provided {provided}.')
                self.description = desc
            provided.update(segment.gives)
        self.kind = kind
    return self
