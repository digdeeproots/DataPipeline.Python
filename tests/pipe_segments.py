import asyncio
from typing import Any, Generic, TypeVar

import pytest
from assertpy import assert_that
from approvaltests import verify
from datapipeline import RestructuringSegment, SourceSegment, TransformSegment, SinkSegment, PipeHeadSegment


T = TypeVar('T')


class DataTransferObject:
    some_value: Any
    dumb_object: Any

    def __init__(self, some_value=None):
        self.some_value = some_value
        self.dumb_object = None


class SecondDTO:
    message: str

    def __init__(self, message: str = ""):
        self.message = message


def format_dumb_object(data: DataTransferObject) -> None:
    data.dumb_object = f"{{ saw: {data.some_value} }}"


async def first(data: DataTransferObject) -> dict:
    pass


def parse_first(data: DataTransferObject, new_data: dict) -> None:
    pass


def second(data: DataTransferObject) -> None:
    pass


def extract_for_third(data: SecondDTO) -> dict:
    pass


async def third(data: dict) -> None:
    pass


def convert(data: DataTransferObject) -> SecondDTO:
    return SecondDTO(f"made by convert function from {data.some_value}")


class Capture(Generic[T]):
    result: T

    def __init__(self):
        self.result = None
        self.__name__ = 'Capture'

    def __call__(self, arg: T) -> None:
        self.result = arg


class CaptureAsync(Generic[T]):
    result: T

    def __init__(self):
        self.result = None
        self.__name__ = 'CaptureAsync'

    async def __call__(self, arg: T) -> None:
        await asyncio.sleep(0.005)
        self.result = arg


async def test_restructuring_segment_uses_its_impl():
    def has_right_message(arg: SecondDTO):
        assert_that(arg.message).is_equal_to("made by convert function from 2")

    test_subject = RestructuringSegment(convert, TransformSegment(has_right_message))
    await test_subject.process(DataTransferObject(2))


async def test_pipe_segment_calls_its_transform_impl_to_process_data():
    capture = Capture[DataTransferObject]()
    test_subject = TransformSegment(format_dumb_object, TransformSegment(capture))
    await test_subject.process(DataTransferObject(4))
    assert_that(capture.result.dumb_object).is_equal_to("{ saw: 4 }")


async def test_source_segment_chains_its_two_implementations():
    async def fetch_data(arg: DataTransferObject) -> str:
        await asyncio.sleep(0.005)
        return "fetched value"

    def parse_data(arg: DataTransferObject, new_data: str) -> None:
        arg.some_value = new_data

    capture = Capture[DataTransferObject]()
    test_subject = SourceSegment(fetch_data, parse_data, TransformSegment(capture))
    await test_subject.process(DataTransferObject("original value"))
    assert_that(capture.result.some_value).is_equal_to("fetched value")


async def test_sink_segment_chains_its_two_implementations():
    def extract(arg: DataTransferObject) -> str:
        return "extracted value"

    capture = CaptureAsync[DataTransferObject]()
    test_subject = SinkSegment(extract, capture)
    await test_subject.process(DataTransferObject("original value"))
    assert_that(capture.result).is_equal_to("extracted value")


def test_pipelines_can_be_validated_as_strings():
    test_subject = PipeHeadSegment(
        DataTransferObject,
        SourceSegment(
            first,
            parse_first,
            TransformSegment(
                second,
                RestructuringSegment(
                    convert,
                    SinkSegment(
                        extract_for_third,
                        third)))))
    verify(test_subject.to_verification_string())
