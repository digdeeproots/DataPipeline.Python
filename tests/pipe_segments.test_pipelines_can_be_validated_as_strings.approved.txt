 <!> Start with empty <DataTransferObject>
  |
>-|   load:first, parse: parse_first
  |     info change: [] ==> []
  |
  +   second
  |     info change: [] ==> []
  |
 <->  Changed to <SecondDTO> using convert
  |     info change: [] ==> []
  |
  |-> extract: extract_for_third, store: third
  |     info change: [] ==> []
